File: old_bak/golang/junior/子集.go
package main

import "fmt"

func subsets(nums []int) [][]int {
	l := len(nums)
	if l == 1 {
		return [][]int{
			{},
			{nums[0]},
		}
	}

	subs := subsets(nums[1:])
	res := [][]int{}
	for _, v := range subs {
		nv := append(v, nums[0])

		res = append(res, append(v, nums[0]))
		res = append(res, v)
	}

	return res
}

func main() {
	nums := []int{1, 2, 3, 4, 5}
	fmt.Println("len: %v\n", len(subsets(nums)))
	fmt.Println(subsets(nums))
}

File: old_bak/python/回溯算法/子集.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        def dfs(nums):
            if len(nums) == 1:
                return [nums]

            res = []

            tmp_res = dfs(nums[1:])

            res.append([nums[0]])
            for item in tmp_res:
                res.append([nums[0]] + item)

            res = res + tmp_res

            return res

        res = dfs(nums)

        return res+[[]]

nums = [1, 2, 3, 4]
print(Solution().subsets(nums))